auto_archive.py

```diff
@@ -58,169 +44,101 @@ def batch_if_valid(col, val, final_value=None):
-                active_archivers = [
-                    archivers.TelethonArchiver(storage_client, driver, telegram_config),
-                    archivers.TelegramArchiver(storage_client, driver),
-                    archivers.TiktokArchiver(storage_client, driver),
-                    archivers.YoutubeDLArchiver(storage_client, driver, os.getenv('FACEBOOK_COOKIE')),
-                    archivers.TwitterArchiver(storage_client, driver),
-                    archivers.WaybackArchiver(storage_client, driver)
-                ]
-                for archiver in active_archivers:
-                    logger.debug(f'Trying {archiver} on row {row}')
-
-                    try:
-                        result = archiver.download(url, check_if_exists=True)
-                    except KeyboardInterrupt:
-                        logger.warning("caught interrupt")
-                        gw.set_cell(row, 'status', '')
-                        driver.quit()
-                        exit()
-                    except Exception as e:
-                        result = False
-                        logger.error(f'Got unexpected error in row {row} with archiver {archiver} for url {url}: {e}\n{traceback.format_exc()}')
-
-                    if result:
-                        # IA is a Success I believe - or do we want to display a logger warning for it?
-                        if result.status in ['success', 'already archived', 'Internet Archive fallback']:
-                            result.status = archiver.name + \
-                                ": " + str(result.status)
-                            logger.success(
-                                f'{archiver} succeeded on row {row}, url {url}')
-                            break
-
-                        # wayback has seen this url before so keep existing status
-                        if "wayback: Internet Archive fallback" in result.status:
-                            logger.success(
-                                f'wayback has seen this url before so keep existing status on row {row}')
-                            result.status = result.status.replace(' (duplicate)', '')
-                            result.status = str(result.status) + " (duplicate)"
-                            break
-
-                        logger.warning(
-                            f'{archiver} did not succeed on {row=}, final status: {result.status}')
-                        result.status = archiver.name + \
-                            ": " + str(result.status)
-                # get rid of driver so can reload on next row
-                driver.quit()
+            is_retry = False
+            if url == '' or status not in ['', None]:
+                is_retry = Archiver.should_retry_from_status(status)
+                if not is_retry: continue
+
+            # All checks done - archival process starts here
+            gw.set_cell(row, 'status', 'Archive in progress')
```

Comment:
An exception here can result in the sheet being left with a status of "archive in progress" when the auto archiver has crashed. Exceptions in this area are likely the result of a configuration error, but I think it would still be best to avoid leaving things in the sheet. Example (from not having geckodriver in the path)

```
2022-06-10 11:22:52.678 | INFO     | __main__:process_sheet:61 - Opening worksheet ii=0: wks.title='Sheet1' c.header=1
2022-06-10 11:22:53.619 | ERROR    | __main__:<module>:145 - An error has been caught in function '<module>', process 'MainProcess' (69460), thread 'MainThread' (4340336000):
Traceback (most recent call last):

  File "/Users/loganw/.local/share/virtualenvs/auto-archiver-89gA8FnB/lib/python3.9/site-packages/selenium/webdriver/common/service.py", line 71, in start
    self.process = subprocess.Popen(cmd, env=self.env,
    │              │          │     │        │    └ environ({'PWD': '/Volumes/Oxygen/auto-archiver', 'COLORTERM': 'truecolor', 'LOGNAME': 'loganw', 'USER': 'loganw', 'ITERM_PROF...
    │              │          │     │        └ <selenium.webdriver.firefox.service.Service object at 0x109306bb0>
    │              │          │     └ ['geckodriver', '--websocket-port', '56069', '--port', '56068']
    │              │          └ <class 'subprocess.Popen'>
    │              └ <module 'subprocess' from '/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/su...
    └ <selenium.webdriver.firefox.service.Service object at 0x109306bb0>

  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
    │    │              │     │           │           └ True
    │    │              │     │           └ None
    │    │              │     └ None
    │    │              └ ['geckodriver', '--websocket-port', '56069', '--port', '56068']
    │    └ <function Popen._execute_child at 0x10336a310>
    └ <Popen: returncode: 255 args: ['geckodriver', '--websocket-port', '56069', '...>

  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 1821, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
          │                    │          │        └ 'geckodriver'
          │                    │          └ 'No such file or directory'
          │                    └ 2
          └ <class 'OSError'>

FileNotFoundError: [Errno 2] No such file or directory: 'geckodriver'


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

> File "/Volumes/Oxygen/auto-archiver/auto_archive.py", line 145, in <module>
    main()
    └ <function main at 0x1092c60d0>

  File "/Volumes/Oxygen/auto-archiver/auto_archive.py", line 139, in main
    process_sheet(c)
    │             └ <configs.config.Config object at 0x103968100>
    └ <function process_sheet at 0x109246f70>

  File "/Volumes/Oxygen/auto-archiver/auto_archive.py", line 88, in process_sheet
    c.recreate_webdriver()
    │ └ <function Config.recreate_webdriver at 0x10850b670>
    └ <configs.config.Config object at 0x103968100>

  File "/Volumes/Oxygen/auto-archiver/configs/config.py", line 208, in recreate_webdriver
    self.webdriver = webdriver.Firefox(options=options)
    │    │           │         │               └ <selenium.webdriver.firefox.options.Options object at 0x109306af0>
    │    │           │         └ <class 'selenium.webdriver.firefox.webdriver.WebDriver'>
    │    │           └ <module 'selenium.webdriver' from '/Users/loganw/.local/share/virtualenvs/auto-archiver-89gA8FnB/lib/python3.9/site-packages/...
    │    └ 'not initialized'
    └ <configs.config.Config object at 0x103968100>

  File "/Users/loganw/.local/share/virtualenvs/auto-archiver-89gA8FnB/lib/python3.9/site-packages/selenium/webdriver/firefox/webdriver.py", line 172, in __init__
    self.service.start()
    │    │       └ <function Service.start at 0x1044b6310>
    │    └ <selenium.webdriver.firefox.service.Service object at 0x109306bb0>
    └ <unprintable WebDriver object>
  File "/Users/loganw/.local/share/virtualenvs/auto-archiver-89gA8FnB/lib/python3.9/site-packages/selenium/webdriver/common/service.py", line 81, in start
    raise WebDriverException(
          └ <class 'selenium.common.exceptions.WebDriverException'>

selenium.common.exceptions.WebDriverException: Message: 'geckodriver' executable needs to be in PATH.
```